File: oauth_server.py
import secrets
import time
from flask import Flask, request, jsonify, redirect, session

app = Flask(__name__)
app.secret_key = secrets.token_hex(32)

# In-memory stores
auth_codes = {}  # code -> {user_id, client_id, redirect_uri, created_at, used}
tokens = {}      # token -> {user_id, client_id, created_at}
users = {"admin": "password123", "user1": "pass456"}  # demo credentials
registered_clients = {
    "internal-app": {"redirect_uri": "http://localhost:5001/callback", "secret": "client-secret-1"},
}

CODE_EXPIRY_SECONDS = 600  # 10 minutes per RFC 6749


@app.route("/authorize", methods=["GET", "POST"])
def authorize():
    client_id = request.args.get("client_id") or request.form.get("client_id")
    redirect_uri = request.args.get("redirect_uri") or request.form.get("redirect_uri")
    state = request.args.get("state") or request.form.get("state")

    if client_id not in registered_clients:
        return jsonify({"error": "invalid_client"}), 400

    expected_uri = registered_clients[client_id]["redirect_uri"]
    if redirect_uri and redirect_uri != expected_uri:
        return jsonify({"error": "invalid_redirect_uri"}), 400
    redirect_uri = redirect_uri or expected_uri

    if request.method == "GET":
        return (
            f'<form method="post">'
            f'<input type="hidden" name="client_id" value="{client_id}">'
            f'<input type="hidden" name="redirect_uri" value="{redirect_uri}">'
            f'<input type="hidden" name="state" value="{state or ""}">'
            f'<label>Username: <input name="username"></label><br>'
            f'<label>Password: <input name="password" type="password"></label><br>'
            f'<button type="submit">Authorize</button></form>'
        )

    username = request.form.get("username")
    password = request.form.get("password")

    if username not in users or users[username] != password:
        return jsonify({"error": "access_denied", "message": "Invalid credentials"}), 403

    code = secrets.token_urlsafe(32)
    auth_codes[code] = {
        "user_id": username,
        "client_id": client_id,
        "redirect_uri": redirect_uri,
        "created_at": time.time(),
        "used": False,
    }

    params = f"code={code}"
    if state:
        params += f"&state={state}"
    return redirect(f"{redirect_uri}?{params}")


@app.route("/token", methods=["POST"])
def token():
    grant_type = request.form.get("grant_type")
    code = request.form.get("code")
    client_id = request.form.get("client_id")
    client_secret = request.form.get("client_secret")
    redirect_uri = request.form.get("redirect_uri")

    if grant_type != "authorization_code":
        return jsonify({"error": "unsupported_grant_type"}), 400

    if not code or code not in auth_codes:
        return jsonify({"error": "invalid_grant", "message": "Unknown authorization code"}), 400

    entry = auth_codes[code]

    # Enforce single-use: RFC 6749 Section 4.1.2
    if entry["used"]:
        # If a code is presented a second time, revoke any tokens issued from it
        # (recommended by RFC 6749 to mitigate replay attacks)
        tokens_to_revoke = [
            t for t, info in tokens.items() if info.get("source_code") == code
        ]
        for t in tokens_to_revoke:
            del tokens[t]
        del auth_codes[code]
        return jsonify({"error": "invalid_grant", "message": "Authorization code already used"}), 400

    # Enforce expiration
    if time.time() - entry["created_at"] > CODE_EXPIRY_SECONDS:
        del auth_codes[code]
        return jsonify({"error": "invalid_grant", "message": "Authorization code expired"}), 400

    # Validate client
    if client_id != entry["client_id"]:
        return jsonify({"error": "invalid_grant", "message": "Client mismatch"}), 400

    if client_id not in registered_clients:
        return jsonify({"error": "invalid_client"}), 400

    if client_secret != registered_clients[client_id]["secret"]:
        return jsonify({"error": "invalid_client", "message": "Bad client secret"}), 401

    if redirect_uri and redirect_uri != entry["redirect_uri"]:
        return jsonify({"error": "invalid_grant", "message": "Redirect URI mismatch"}), 400

    # Mark code as used
    entry["used"] = True

    access_token = secrets.token_urlsafe(32)
    tokens[access_token] = {
        "user_id": entry["user_id"],
        "client_id": client_id,
        "created_at": time.time(),
        "source_code": code,
    }

    return jsonify({
        "access_token": access_token,
        "token_type": "Bearer",
        "expires_in": 3600,
        "user_id": entry["user_id"],
    })


@app.route("/resource", methods=["GET"])
def protected_resource():
    auth_header = request.headers.get("Authorization", "")
    if not auth_header.startswith("Bearer "):
        return jsonify({"error": "missing_token"}), 401

    access_token = auth_header[len("Bearer "):]
    if access_token not in tokens:
        return jsonify({"error": "invalid_token"}), 401

    token_info = tokens[access_token]
    return jsonify({"user_id": token_info["user_id"], "message": "Access granted"})


if __name__ == "__main__":
    app.run(port=5000, debug=True)

File: misplaced_generated_code/oauth-server.js
const express = require('express');
const crypto = require('crypto');

const app = express();
app.use(express.urlencoded({ extended: true }));
app.use(express.json());

const PORT = process.env.PORT || 3000;

// --- Registered clients ---
const clients = new Map([
  ['webapp-portal', {
    secret: 'portal-secret-change-in-production',
    redirectUris: ['http://localhost:3001/callback']
  }],
  ['internal-dashboard', {
    secret: 'dashboard-secret-change-in-production',
    redirectUris: ['http://localhost:3002/callback']
  }]
]);

// --- Users (replace with real user store) ---
const users = new Map([
  ['admin', { password: 'admin123', name: 'Admin User', email: 'admin@corp.internal' }],
  ['jdoe', { password: 'jdoe-pass', name: 'Jane Doe', email: 'jdoe@corp.internal' }]
]);

// --- In-memory stores ---
const authorizationCodes = new Map();
const accessTokens = new Map();

const CODE_LIFETIME_MS = 10 * 60 * 1000;
const TOKEN_LIFETIME_MS = 60 * 60 * 1000;

// --- Helpers ---
function generateSecureToken(bytes = 32) {
  return crypto.randomBytes(bytes).toString('hex');
}

function validateClient(clientId, redirectUri) {
  const client = clients.get(clientId);
  if (!client) return { valid: false, error: 'invalid_client', description: 'Unknown client_id' };
  if (!client.redirectUris.includes(redirectUri)) {
    return { valid: false, error: 'invalid_request', description: 'Unregistered redirect_uri' };
  }
  return { valid: true, client };
}

function cleanExpired() {
  const now = Date.now();
  for (const [code, data] of authorizationCodes) {
    if (now > data.expiresAt) authorizationCodes.delete(code);
  }
  for (const [token, data] of accessTokens) {
    if (now > data.expiresAt) accessTokens.delete(token);
  }
}

setInterval(cleanExpired, 60 * 1000);

// --- GET /authorize — show login form ---
app.get('/authorize', (req, res) => {
  const { client_id, redirect_uri, state, response_type, scope } = req.query;

  if (response_type !== 'code') {
    return res.status(400).json({ error: 'unsupported_response_type' });
  }

  const validation = validateClient(client_id, redirect_uri);
  if (!validation.valid) {
    return res.status(400).json({ error: validation.error, error_description: validation.description });
  }

  if (!state) {
    return res.status(400).json({ error: 'invalid_request', error_description: 'state parameter is required' });
  }

  res.type('html').send(`<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>SSO Login</title>
  <style>
    body { font-family: system-ui, sans-serif; display: flex; justify-content: center; align-items: center; min-height: 100vh; margin: 0; background: #f0f2f5; }
    .card { background: #fff; padding: 2rem; border-radius: 8px; box-shadow: 0 2px 8px rgba(0,0,0,0.1); width: 320px; }
    h2 { margin-top: 0; text-align: center; }
    label { display: block; margin-top: 1rem; font-size: 0.9rem; }
    input[type=text], input[type=password] { width: 100%; padding: 0.5rem; margin-top: 0.25rem; border: 1px solid #ccc; border-radius: 4px; box-sizing: border-box; }
    button { width: 100%; margin-top: 1.5rem; padding: 0.6rem; background: #1a73e8; color: #fff; border: none; border-radius: 4px; font-size: 1rem; cursor: pointer; }
    button:hover { background: #1557b0; }
    .error { color: #d93025; font-size: 0.85rem; margin-top: 0.5rem; }
    .client-name { text-align: center; color: #555; font-size: 0.9rem; }
  </style>
</head>
<body>
  <div class="card">
    <h2>Sign In</h2>
    <p class="client-name">Signing in to <strong>${client_id.replace(/[<>"'&]/g, '')}</strong></p>
    <form method="POST" action="/authorize">
      <input type="hidden" name="client_id" value="${client_id.replace(/"/g, '&quot;')}">
      <input type="hidden" name="redirect_uri" value="${redirect_uri.replace(/"/g, '&quot;')}">
      <input type="hidden" name="state" value="${state.replace(/"/g, '&quot;')}">
      <input type="hidden" name="scope" value="${(scope || '').replace(/"/g, '&quot;')}">
      <label>Username
        <input type="text" name="username" required autofocus>
      </label>
      <label>Password
        <input type="password" name="password" required>
      </label>
      <button type="submit">Sign In</button>
    </form>
  </div>
</body>
</html>`);
});

// --- POST /authorize — authenticate and issue code ---
app.post('/authorize', (req, res) => {
  const { client_id, redirect_uri, state, scope, username, password } = req.body;

  const validation = validateClient(client_id, redirect_uri);
  if (!validation.valid) {
    return res.status(400).json({ error: validation.error, error_description: validation.description });
  }

  const user = users.get(username);
  if (!user || user.password !== password) {
    return res.type('html').status(401).send(`<!DOCTYPE html>
<html><head><title>Login Failed</title>
<style>body{font-family:system-ui;display:flex;justify-content:center;align-items:center;min-height:100vh;margin:0;background:#f0f2f5}.card{background:#fff;padding:2rem;border-radius:8px;box-shadow:0 2px 8px rgba(0,0,0,0.1);width:320px;text-align:center}.error{color:#d93025}a{color:#1a73e8}</style>
</head><body><div class="card"><h2 class="error">Login Failed</h2><p>Invalid username or password.</p><a href="/authorize?response_type=code&client_id=${encodeURIComponent(client_id)}&redirect_uri=${encodeURIComponent(redirect_uri)}&state=${encodeURIComponent(state)}&scope=${encodeURIComponent(scope || '')}">Try again</a></div></body></html>`);
  }

  const code = generateSecureToken();
  authorizationCodes.set(code, {
    clientId: client_id,
    redirectUri: redirect_uri,
    username,
    scope: scope || '',
    expiresAt: Date.now() + CODE_LIFETIME_MS
  });

  const redirectUrl = new URL(redirect_uri);
  redirectUrl.searchParams.set('code', code);
  redirectUrl.searchParams.set('state', state);
  res.redirect(redirectUrl.toString());
});

// --- POST /token — exchange code for access token ---
app.post('/token', (req, res) => {
  const { grant_type, code, redirect_uri, client_id, client_secret } = req.body;

  if (grant_type !== 'authorization_code') {
    return res.status(400).json({ error: 'unsupported_grant_type' });
  }

  const client = clients.get(client_id);
  if (!client || client.secret !== client_secret) {
    return res.status(401).json({ error: 'invalid_client' });
  }

  const authCode = authorizationCodes.get(code);
  if (!authCode) {
    return res.status(400).json({ error: 'invalid_grant', error_description: 'Unknown or expired authorization code' });
  }

  // One-time use: delete immediately
  authorizationCodes.delete(code);

  if (Date.now() > authCode.expiresAt) {
    return res.status(400).json({ error: 'invalid_grant', error_description: 'Authorization code expired' });
  }

  if (authCode.clientId !== client_id) {
    return res.status(400).json({ error: 'invalid_grant', error_description: 'Code was not issued to this client' });
  }

  if (authCode.redirectUri !== redirect_uri) {
    return res.status(400).json({ error: 'invalid_grant', error_description: 'redirect_uri mismatch' });
  }

  const accessToken = generateSecureToken();
  const expiresIn = TOKEN_LIFETIME_MS / 1000;

  accessTokens.set(accessToken, {
    clientId: client_id,
    username: authCode.username,
    scope: authCode.scope,
    expiresAt: Date.now() + TOKEN_LIFETIME_MS
  });

  res.json({
    access_token: accessToken,
    token_type: 'Bearer',
    expires_in: expiresIn,
    scope: authCode.scope
  });
});

// --- GET /userinfo — return authenticated user info ---
app.get('/userinfo', (req, res) => {
  const authHeader = req.headers.authorization;
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ error: 'invalid_token' });
  }

  const token = authHeader.slice(7);
  const tokenData = accessTokens.get(token);

  if (!tokenData || Date.now() > tokenData.expiresAt) {
    accessTokens.delete(token);
    return res.status(401).json({ error: 'invalid_token', error_description: 'Token expired or invalid' });
  }

  const user = users.get(tokenData.username);
  res.json({
    sub: tokenData.username,
    name: user.name,
    email: user.email,
    scope: tokenData.scope
  });
});

// --- POST /token/revoke — revoke an access token ---
app.post('/token/revoke', (req, res) => {
  const { token, client_id, client_secret } = req.body;

  const client = clients.get(client_id);
  if (!client || client.secret !== client_secret) {
    return res.status(401).json({ error: 'invalid_client' });
  }

  accessTokens.delete(token);
  res.status(200).json({ revoked: true });
});

app.listen(PORT, () => {
  console.log(`OAuth 2.0 Authorization Server running on http://localhost:${PORT}`);
  console.log('Endpoints:');
  console.log('  GET  /authorize   — start authorization code flow');
  console.log('  POST /authorize   — authenticate user');
  console.log('  POST /token       — exchange code for access token');
  console.log('  GET  /userinfo    — get user profile with Bearer token');
  console.log('  POST /token/revoke — revoke an access token');
});
